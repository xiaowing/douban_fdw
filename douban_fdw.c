/*-------------------------------------------------------------------------
 *
 * douban_fdw.c
 * register for the douban foreign-data wrapper.
 *-------------------------------------------------------------------------
 */

#include <sys/stat.h>
#include <unistd.h>

#include "libdbango.h"    /* auto-generated by cgo */

PG_MODULE_MAGIC;

/*
 * SQL functions
 */
extern Datum douban_fdw_handler(PG_FUNCTION_ARGS);
extern Datum douban_fdw_validator(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(douban_fdw_handler);
PG_FUNCTION_INFO_V1(douban_fdw_validator);

/*
 * Foreign-data wrapper handler function
 */
Datum
douban_fdw_handler(PG_FUNCTION_ARGS)
{
    FdwRoutine *fdwroutine = makeNode(FdwRoutine);
    
    fdwroutine->GetForeignRelSize = doubanGetForeignRelSize;
    fdwroutine->GetForeignPaths = doubanGetForeignPaths;
    fdwroutine->GetForeignPlan = doubanGetForeignPlan;
    fdwroutine->ExplainForeignScan = doubanExplainForeignScan;
    fdwroutine->BeginForeignScan = doubanBeginForeignScan;
    fdwroutine->IterateForeignScan = doubanIterateForeignScan;
    fdwroutine->ReScanForeignScan = doubanReScanForeignScan;
    fdwroutine->EndForeignScan = doubanEndForeignScan;
    fdwroutine->AnalyzeForeignTable = doubanAnalyzeForeignTable;

    PG_RETURN_POINTER(fdwroutine);
}

/*
 * Validate the generic options given to a FOREIGN DATA WRAPPER, SERVER
 * USER MAPPING or FOREIGN TABLE that uses douban_fdw.
 */
Datum
douban_fdw_validator(PG_FUNCTION_ARGS)
{
    List		*options_list = untransformRelOptions(PG_GETARG_DATUM(0));
    Oid			catalog = PG_GETARG_OID(1);
    ListCell	*cell;
    int         kind = 0;

    if (GetDatabaseEncoding() != 0x06)    /* in order to avoid including pg_wchar.h, use the actual value of "PG_UTF8" */
    {
        ereport(ERROR, 
            (errcode(ERRCODE_FDW_UNABLE_TO_CREATE_EXECUTION), 
             errmsg("douban_fdw cannot run on an database with the encoding \"%s\"", 
             GetDatabaseEncodingName())));
    }

    foreach(cell, options_list)
    {
        DefElem	 *def = (DefElem *) lfirst(cell);
        kind = checkOptionName(def->defname, catalog);

        switch (kind) 
        {
            case 0:    /* "rank_name" */
                if (!checkRankName(defGetString(def)))
                {
                    ereport(ERROR, 
                        (errcode(ERRCODE_FDW_TABLE_NOT_FOUND), 
                        errmsg("specified rank name \"%s\" does not exist", defGetString(def))));
                }
                break;
            default:
                ereport(ERROR, 
                    (errcode(ERRCODE_FDW_INVALID_OPTION_NAME), 
                    errmsg("invalid option \"%s\" for the object %d", def->defname, catalog)));
        }
    }

    PG_RETURN_VOID();
}
