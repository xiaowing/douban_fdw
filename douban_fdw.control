# douban_fdw extension
comment = 'foreign-data wrapper for douban public API'
default_version = '1.0'
module_pathname = '$libdir/douban_fdw'
relocatable = true
