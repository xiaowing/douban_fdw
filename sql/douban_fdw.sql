CREATE EXTENSION douban_fdw;
CREATE SERVER doubansv FOREIGN DATA WRAPPER douban_fdw;
CREATE FOREIGN TABLE top250(rating REAL, title TEXT, genres VARCHAR(64), casts VARCHAR(256), collectcount INTEGER, originname TEXT, directors VARCHAR(256), year VARCHAR(32)) SERVER doubansv OPTIONS(rank_name 'top250');

SELECT count(1) FROM top250;

SELECT originname, title FROM top250 WHERE CAST(year AS INT) = 1994 ORDER BY rating DESC, title;

SELECT title, year FROM top250 WHERE rating NOT IN (SELECT rating FROM top250 WHERE year::int > 2000) ORDER BY rating DESC, title;

EXPLAIN SELECT rating, title, year FROM top250 WHERE rating NOT IN (SELECT rating FROM top250 WHERE year::int > 2000) ORDER BY rating DESC;

DROP FOREIGN TABLE top250;
DROP SERVER doubansv;
DROP EXTENSION douban_fdw CASCADE;