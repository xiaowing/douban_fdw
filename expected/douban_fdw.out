CREATE EXTENSION douban_fdw;
CREATE SERVER doubansv FOREIGN DATA WRAPPER douban_fdw;
CREATE FOREIGN TABLE top250(rating REAL, title TEXT, genres VARCHAR(64), casts VARCHAR(256), collectcount INTEGER, originname TEXT, directors VARCHAR(256), year VARCHAR(32)) SERVER doubansv OPTIONS(rank_name 'top250');
CREATE FOREIGN TABLE usbox(rating REAL, title TEXT, genres VARCHAR(64), casts VARCHAR(256), collectcount INTEGER, originname TEXT, directors VARCHAR(256), year VARCHAR(32)) SERVER doubansv OPTIONS(rank_name 'us_box');
SELECT count(1) FROM top250;
 count 
-------
   250
(1 row)

SELECT count(*) FROM usbox;
 count 
-------
    10
(1 row)

SELECT originname, title FROM top250 WHERE CAST(year AS INT) = 1994 ORDER BY rating DESC, title;
        originname        |     title      
--------------------------+----------------
 The Shawshank Redemption | 肖申克的救赎
 Léon                     | 这个杀手不太冷
 Forrest Gump             | 阿甘正传
 活着                     | 活着
 飲食男女                 | 饮食男女
 The Lion King            | 狮子王
 Pulp Fiction             | 低俗小说
 Legends of the Fall      | 燃情岁月
 阳光灿烂的日子           | 阳光灿烂的日子
 重慶森林                 | 重庆森林
 東邪西毒                 | 东邪西毒
(11 rows)

EXPLAIN SELECT rating, title, year FROM top250 WHERE rating NOT IN (SELECT rating FROM top250 WHERE year::int > 2000) ORDER BY rating DESC;
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 Sort  (cost=299.96..300.58 rows=250 width=118)
   Sort Key: top250.rating DESC
   ->  Foreign Scan on top250  (cost=40.00..290.00 rows=250 width=118)
         Filter: (NOT (hashed SubPlan 1))
         Douban Rank: top250
         Movie items: 250
         SubPlan 1
           ->  Foreign Scan on top250 top250_1  (cost=40.00..290.00 rows=250 width=4)
                 Filter: ((year)::integer > 2000)
                 Douban Rank: top250
                 Movie items: 250
(11 rows)

DROP FOREIGN TABLE usbox;
DROP FOREIGN TABLE top250;
DROP SERVER doubansv;
DROP EXTENSION douban_fdw CASCADE;
